eventstreamrpc/include/aws/eventstreamrpc/EventStreamClient.h

```diff
@@ -0,0 +1,278 @@
+         * Invoked upon receiving any error. Use the return value to determine
+         * whether or not to force the connection to close. Keep in mind that once
+         * closed, the pointer reference from `OnConnect` is no longer safe to use.
+         */
+        using OnError = std::function<bool(int errorCode)>;
+
+        /**
+         * Invoked upon receiving a ping from the server. The `headers` and `payload`
+         * refer to what is contained in the ping message.
+         */
+        using OnPing = std::function<
+            void(const Crt::List<EventstreamHeader> &headers, const Crt::Optional<Crt::ByteBuf> &payload)>;
+
+        /**
+         * Allows the application to append headers and change the payload of the CONNECT
+         * packet being sent out.
+         */
+        using ConnectMessageAmender = std::function<MessageAmendment &(void)>;
+
+        class AWS_EVENTSTREAMRPC_API EventstreamHeader final
+        {
+          public:
+            EventstreamHeader(const EventstreamHeader &lhs) noexcept;
+            EventstreamHeader(EventstreamHeader &&rhs) noexcept;
+            ~EventstreamHeader() noexcept;
+            EventstreamHeader(const struct aws_event_stream_header_value_pair &header);
+            EventstreamHeader(const Crt::String &name, bool value);
+            EventstreamHeader(const Crt::String &name, int8_t value);
+            EventstreamHeader(const Crt::String &name, int16_t value);
+            EventstreamHeader(const Crt::String &name, int32_t value);
+            EventstreamHeader(const Crt::String &name, int64_t value);
+            EventstreamHeader(const Crt::String &name, Crt::DateTime &value);
+            EventstreamHeader(
+                const Crt::String &name,
+                const Crt::String &value,
+                Crt::Allocator *allocator = Crt::g_allocator) noexcept;
+            EventstreamHeader(const Crt::String &name, Crt::ByteBuf &value);
+            EventstreamHeader(const Crt::String &name, Crt::UUID value);
+
+            HeaderType GetHeaderType();
+            Crt::String GetHeaderName() noexcept;
+            void SetHeaderName(Crt::String &);
+
+            bool GetValueAsBoolean(bool &);
+            bool GetValueAsByte(int8_t &);
+            bool GetValueAsShort(int16_t &);
+            bool GetValueAsInt(int32_t &);
+            bool GetValueAsLong(int64_t &);
+            bool GetValueAsTimestamp(Crt::DateTime &);
+            bool GetValueAsBytes(Crt::ByteBuf &);
+            bool GetValueAsUUID(Crt::UUID &);
+
+            void SetValue(bool value);
+            void SetValue(int8_t value);
+            void SetValue(int16_t value);
+            void SetValue(int32_t value);
+            void SetValue(int64_t value);
+            void SetValue(Crt::DateTime &value);
+            void SetValue(Crt::ByteBuf value);
+            void SetValue(Crt::UUID value);
+
+            const struct aws_event_stream_header_value_pair *GetUnderlyingHandle() const;
```

Comment:
The useful function that you want is to go from a C++ `EventStreamHeader` to the C `aws_event_stream_header_value_pair`. To do this, write an explicit conversion operator member function:

`explicit operator aws_event_stream_header_value_pair() const`

I'm fairly certain C++11 rules suggest return value optimization will occur on C structs so write it naively. That is return the stack allocated aws_event_stream_header_value_pair created inside the conversion by value. Don't try to move it or anything in the return statement.